### PR TITLE
Add softmaxV11ToLatest for non-innermost axis

### PR DESCRIPTION
**Pull request by tr00**
This PR improves the pre-existing canonicalization pattern which rewrites SoftmaxV11 to Softmax. They have different semantics and the old pattern only converted in case the dimension was last/innermost. 
The new pattern reshapes the input tensor before and after such that the new version emulates the old one.

Example:
`onnx.SoftmaxV11(%x : tensor<1x2x3x4x5xf32>) { axis = 2} `
becomes
```
onnx.Reshape(...) : tensor<1x2x3x4x5xf32> -> tensor<1x2x60xf32>
onnx.Softmax(...) { axis = 2 } : tensor<1x2x60xf32> -> tensor<1x2x60xf32>
onnx.Reshape(...) : tensor<1x2x60xf32> -> tensor<1x2x3x4x5xf32>
```

See [link](https://onnx.ai/onnx/operators/onnx__Softmax.html#softmax-11) for the docs of SoftmaxV11

### CODE REVIEW

File: src/Dialect/ONNX/ONNXOps/Canonicalize.cpp [hunk 1]
```diff
@@ -2433,6 +2433,76 @@ struct SoftmaxNegativeAxisPattern : public OpRewritePattern<ONNXSoftmaxOp> {
...
+
+    Location loc = op.getLoc();
+    auto shapeAttr = [&](ArrayRef<int64_t> shape) {
+      return rewriter.create<ONNXConstantOp>(
+          loc, nullptr, rewriter.getI64TensorAttr(shape));
+    };
+
+    Value flat = rewriter.create<ONNXReshapeOp>(
+        loc, flatType, input, shapeAttr(flatShape));
+    Value softmax = rewriter.create<ONNXSoftmaxOp>(loc, flatType, flat, axis);
```
Comment (p-lanza) [inline src/Dialect/ONNX/ONNXOps/Canonicalize.cpp]: did you look into using the OnnxBuilder API? Might simplify this code a bit and make it more readable

**Review by p-lanza — APPROVED** [review]
thanks

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 2]
```diff
@@ -1536,6 +1536,40 @@ func.func @test_softmax_v11_unranked(%arg0 : tensor<*xf32>) -> tensor<*xf32> {
...
+// CHECK-LABEL:  func.func @test_softmax_v11_axis_zero
+// CHECK-SAME:   ([[PARAM_0_:%.+]]: tensor<10x20x30xf32>) -> tensor<10x20x30xf32> {
+// CHECK-DAG:       [[SHAPE_ORIG_:%.+]] = onnx.Constant dense<[10, 20, 30]> : tensor<3xi64>
+// CHECK-DAG:       [[SHAPE_FLAT_:%.+]] = onnx.Constant dense<6000> : tensor<1xi64>
+// CHECK:           [[FLAT_:%.+]] = "onnx.Reshape"([[PARAM_0_]], [[SHAPE_FLAT_]]) {{.*}} : (tensor<10x20x30xf32>, tensor<1xi64>) -> tensor<6000xf32>
+// CHECK:           [[SOFTMAX_:%.+]] = "onnx.Softmax"([[FLAT_]]) {axis = 0 : si64} : (tensor<6000xf32>) -> tensor<6000xf32>
+// CHECK:           [[RESULT_:%.+]] = "onnx.Reshape"([[SOFTMAX_]], [[SHAPE_ORIG_]]) {{.*}} : (tensor<6000xf32>, tensor<3xi64>) -> tensor<10x20x30xf32>
+// CHECK:           onnx.Return [[RESULT_]] : tensor<10x20x30xf32>
+// CHECK:         }
+}
```
Comment (gerion-amd) [inline test/mlir/onnx/onnx_canonicalization.mlir:1624]: Please also test SofmaxV11 with a axis = -1 (for unranked input?) and a test for another negative value to test those both branches.
Can you also add a negative test for dynamic shapes?

Comment (tr00) [reply test/mlir/onnx/onnx_canonicalization.mlir:1624]: Yes! I've added cases for that now.

File: test/mlir/onnx/onnx_canonicalization.mlir [hunk 3]
```diff
@@ -1536,6 +1536,48 @@ func.func @test_softmax_v11_unranked(%arg0 : tensor<*xf32>) -> tensor<*xf32> {
...
+// CHECK:           [[SOFTMAX_:%.+]] = "onnx.Softmax"([[FLAT_]]) {axis = 1 : si64} : (tensor<10x600xf32>) -> tensor<10x600xf32>
+// CHECK:           [[RESULT_:%.+]] = "onnx.Reshape"([[SOFTMAX_]], [[SHAPE_ORIG_]]) {{.*}} : (tensor<10x600xf32>, tensor<3xi64>) -> tensor<10x20x30xf32>
+// CHECK:           onnx.Return [[RESULT_]] : tensor<10x20x30xf32>
+// CHECK:         }
+}
+
+// -----
+
+func.func @test_softmax_v11_dynamic_shape_unchanged(%arg0 : tensor<12x4x56x?xf32>) -> tensor<12x4x56x?xf32> {
+  %0 = "onnx.SoftmaxV11"(%arg0) {axis = -2 : si64} : (tensor<12x4x56x?xf32>) -> tensor<12x4x56x?xf32>
```
Comment (gerion-amd) [inline test/mlir/onnx/onnx_canonicalization.mlir:1583]: Nit: With axis = -1 this would be supported, wouldn't it? Maybe also add a test for that?

**Review by gerion-amd — APPROVED** [review]
Thanks
